### DISCUSSION TITLE
Want to add Custom NextPageContext type!

### DISCUSSION THREAD

**Discussion by nihaltowfiq** [vercel/next.js#34959] [Help]
IN Next JS version: 11, I add a custom NextPageContext type for typescript!
following this three step:
![0](https://user-images.githubusercontent.com/62647867/156305569-11a55942-2ee7-4837-8a76-da54e0679665.png)
![1st](https://user-images.githubusercontent.com/62647867/156305587-49dc4cd9-857c-4778-9a8c-70981a555542.png)
![2](https://user-images.githubusercontent.com/62647867/156305597-43ccf25d-0537-44dd-bdbb-61132945735e.png)

**BUT** 
in Next JS version: 12, I want to add same custom NextPageContext type for typescript; 
Bt is throw a ERROR 'Property 'isMobile' does not exist on type 'NextPageContext'.ts(2339)';
so how can i add Custom type in NextPageContext??

**Comment by andremalveira** [comment]
Hi! You can try using it this way, it probably fixes the typing problem, but it doesn't guarantee that the boolean value is sent to this property within the context.

```ts
// next.d.ts

declare module 'next' {
   interface NextPageContext {
     isMobile: boolean;
   }
}
```

> I was testing this way, but passing a user session, and when I tried to get this property on other pages, it came as undefined.

The solution I got was to pass the properties I would use in the response (res), like this:

```tsx
 // _app.tsx

 MyApp.getInitialProps = async ({ ctx }:AppContext) => {

    ctx.res.isMobile = isMobileDevice.......
    ......
    .....
 })

```
```ts
// next.d.ts

declare module 'next/app' {
  type AppContext<R extends NextRouter = NextRouter> = {
    Component: NextComponentType<NextPageContext>;
    AppTree: AppTreeType;
    ctx: NextPageContext & {
      res: ServerResponse & {
        isMobile: boolean;
      }
    } ;
    router: R;
  };

}
```
